File: pixels-retina/src/main/java/io/pixelsdb/pixels/retina/Retina.java
/*
 * Copyright 2024 PixelsDB.
 *
 * This file is part of Pixels.
 *
 * Pixels is free software: you can redistribute it and/or modify
 * it under the terms of the Affero GNU General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Pixels is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * Affero GNU General Public License for more details.
 *
 * You should have received a copy of the Affero GNU General Public
 * License along with Pixels.  If not, see
 * <https://www.gnu.org/licenses/>.
 */
package io.pixelsdb.pixels.retina;

import com.sun.jna.Platform;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.File;
import java.nio.file.Paths;

/**
 * This class is used to manage the visibility of a row-group.
 * It provides methods to add visibility, delete record, and garbage collect.
 */
public class Retina implements AutoCloseable
{
    private static final Logger logger = LogManager.getLogger(Retina.class);
    static
    {
        String pixelsHome = System.getenv("PIXELS_HOME");
        if (pixelsHome == null || pixelsHome.isEmpty()) {
            throw new IllegalStateException("Environment variable PIXELS_HOME is not set");
        }

        if (!Platform.isLinux()) {
            logger.error("direct io is not supported on OS other than Linux");
        }
        String libPath = Paths.get(pixelsHome, "lib/libpixels-retina.so").toString();
        File libFile = new File(libPath);
        if (!libFile.exists()) {
            throw new IllegalStateException("libpixels-retina.so not found at " + libPath);
        }
        if (!libFile.canRead()) {
            throw new IllegalStateException("libpixels-retina.so is not readable at " + libPath);
        }
        System.load(libPath);
    }
    
    /**
     * Constructor creates C++ object and returns handle
     */
    private final long nativeHandle;

    public Retina(long rgRecordNum)
    {
        this.nativeHandle = createNativeObject(rgRecordNum);
    }

    @Override
    public void close()
    {
        if (this.nativeHandle != 0) {
            destroyNativeObject(this.nativeHandle);
        }
    }
    
    // native methods
    private native long createNativeObject(long rgRecordNum);
    private native void destroyNativeObject(long nativeHandle);
    private native void beginRowGroupRead(long nativeHandle);
    private native void endRowGroupRead(long nativeHandle);
    private native void deleteRecord(long rowId, long timestamp, long nativeHandle);
    private native void getVisibilityBitmap(long timestamp, long[] visibilityBitmap, long nativeHandle);
    private native void beginGarbageCollect(long timestamp, long nativeHandle);
    private native void endGarbageCollect(long nativeHandle);

    public void beginRowGroupRead()
    {
        beginRowGroupRead(this.nativeHandle);
    }

    public void endRowGroupRead()
    {
        endRowGroupRead(this.nativeHandle);
    }

    public void deleteRecord(long rowId, long timestamp)
    {
        deleteRecord(rowId, timestamp, this.nativeHandle);
    }

    public void getVisibilityBitmap(long timestamp, long[] visibilityBitmap)
    {
        getVisibilityBitmap(timestamp, visibilityBitmap, this.nativeHandle);
    }

    public void beginGarbageCollect(long timestamp)
    {
        beginGarbageCollect(timestamp, this.nativeHandle);
    }

    public void endGarbageCollect()
    {
        endGarbageCollect(this.nativeHandle);
    }
}

File: cpp/pixels-retina/lib/Retina.cpp
/*
 * Copyright 2025 PixelsDB.
 *
 * This file is part of Pixels.
 *
 * Pixels is free software: you can redistribute it and/or modify
 * it under the terms of the Affero GNU General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Pixels is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * Affero GNU General Public License for more details.
 *
 * You should have received a copy of the Affero GNU General Public
 * License along with Pixels.  If not, see
 * <https://www.gnu.org/licenses/>.
 */
#include "Retina.h"
#include <stdexcept>
#include <cstring>
#include <thread>

Retina::Retina(uint64_t rgRecordNum)
    : numVisibilities((rgRecordNum + VISIBILITY_RECORD_CAPACITY - 1) / VISIBILITY_RECORD_CAPACITY) {
    flag.store(0);
    visibilities = new Visibility[numVisibilities];
}

Retina::~Retina() {
    delete[] visibilities;
}

void Retina::beginAccess() {
    while (true) {
        uint32_t v = flag.load(std::memory_order_acquire);
        uint32_t accessCount = v & ACCESS_MASK;

        if (accessCount >= MAX_ACCESS_COUNT) {
            throw std::runtime_error("Reaches the max concurrent access count.");
        }

        if ((v & GC_MASK) > 0 ||
            !flag.compare_exchange_strong(v, v + ACCESS_INC, std::memory_order_acq_rel)) {
            // We failed to get gc lock or increase access count.
            if ((v & GC_MASK) > 0) {
                // if there is an existing gc, sleep for 10ms.
                std::this_thread::sleep_for(std::chrono::milliseconds(10));
            }
            continue;
        }
        break;
    }
}

void Retina::endAccess() {
    uint32_t v = flag.load(std::memory_order_acquire);
    while((v & ACCESS_MASK) > 0) {
        if (flag.compare_exchange_strong(v, v - ACCESS_INC, std::memory_order_acq_rel)) {
            break;
        }
        v = flag.load(std::memory_order_acquire);
    }
}

void Retina::garbageCollect(uint64_t timestamp) {
    // Set the gc flag.
    flag.store(flag.load(std::memory_order_acquire) | GC_MASK, std::memory_order_release);

    // Wait for all access to end.
    while (true) {
        uint32_t v = flag.load(std::memory_order_acquire);
        if ((v & ACCESS_MASK) == 0) {
            break;
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }

    assert((flag.load(std::memory_order_acquire) & GC_MASK) > 0);
    assert((flag.load(std::memory_order_acquire) & ACCESS_MASK) == 0);

    // Garbage collect.
    for (uint64_t i = 0; i < numVisibilities; i++) {
        visibilities[i].garbageCollect(timestamp);
    }

    // Clear the gc flag.
    flag.store(flag.load(std::memory_order_acquire) & ~GC_MASK, std::memory_order_release);
}

Visibility* Retina::getVisibility(uint64_t rowId) const {
    uint64_t visibilityIndex = rowId / VISIBILITY_RECORD_CAPACITY;
    if (visibilityIndex >= numVisibilities) {
        throw std::runtime_error("Row id is out of range.");
    }
    return &visibilities[visibilityIndex];
}

void Retina::deleteRecord(uint64_t rowId, uint64_t timestamp) {
    try {
        // beginAccess();
        Visibility* visibility = getVisibility(rowId);
        visibility->deleteRecord(rowId % VISIBILITY_RECORD_CAPACITY, timestamp);
        // endAccess();
    }
    catch (const std::runtime_error& e) {
        // endAccess();
        throw std::runtime_error("Failed to delete record: " + std::string(e.what()));
    }
}

uint64_t* Retina::getVisibilityBitmap(uint64_t timestamp) {
    // beginAccess();
    uint64_t* bitmap = new uint64_t[numVisibilities * BITMAP_SIZE_PER_VISIBILITY];
    memset(bitmap, 0, numVisibilities * BITMAP_SIZE_PER_VISIBILITY * sizeof(uint64_t));

    try {
        for (uint64_t i = 0; i < numVisibilities; i++) {
            visibilities[i].getVisibilityBitmap(timestamp, bitmap + i * BITMAP_SIZE_PER_VISIBILITY);
        }
        // endAccess();
        return bitmap;
    } catch (const std::runtime_error& e) {
        delete[] bitmap;
        // endAccess();
        throw std::runtime_error("Failed to get visibility bitmap: " + std::string(e.what()));
    }
}
